src/main/scala/format/pgn/Node.scala

```diff
@@ -0,0 +1,101 @@
+package chess
+package format
+package pgn
+
+import cats.syntax.all.*
+import cats.*
+
+import monocle.{ Optional, Traversal }
+
+/**
+ * Node is a tree structure specialized for chess games.
+ * It's a tree with a main line and variations.
+ * Node is a [Functor](https://typelevel.org/cats/typeclasses/functor.html)
+ * Node is a [Traverse](https://typelevel.org/cats/typeclasses/traverse.html)
+ * It is also provided some Monocle optics see Node.filterOptional and Node.filterTraversal
+  */
+case class Node[A](
+    value: A,
+    child: Option[Node[A]],   // main line next move
+    variations: List[Node[A]] // alternate moves, e.g. Node("e4", None, List("d4", "Nf3"))
+):
+  def mainLine: List[A] = value :: child.fold(Nil)(_.mainLine)
+  def totalNodes: Int   = this.foldLeft(0)((b, _) => b + 1)
+
+  // find the first Node that statisfies the predicate
+  def findNode(predicate: A => Boolean): Option[Node[A]] =
+    if predicate(value) then this.some
+    else
+      child
+        .flatMap(_.findNode(predicate))
+        .orElse(variations.foldLeft(none[Node[A]])((b, n) => b.orElse(n.findNode(predicate))))
```

Comment:
yeah, public api for `children` would be nice indeed.

Reply:
That seems strange to me. `variations` are not `children`, they're at the same ply as the the current `Node`.

`Node("1. e4", None, List("1. d4", "1. Nf3"))`

To me it's unintuitive to call `1 d4` and `1 Nf3` "children" of `1. e4`.

So maybe another word?

Reply:
Good catch I was confused about this and was thinking in terms of the old structure.

Reply:
This is what I would expect `children: List[Node[A]]` to be as an API user:

```scala
  def children = child.fold(Nil)(c => c :: c.variations)
```

That's because the way I visualize a move tree is not the way it is encoded here. There's a shift that could cause accidents if we're not careful with naming.
So maybe this:

```scala
  def mainLineAndVariations = child.map(_ :: variations).getOrElse(variations)
  def children = child.fold(Nil)(c => c :: c.variations)
```